extmod/moducryptolib.c

```diff
@@ -356,9 +361,190 @@ STATIC const mp_obj_type_t ucryptolib_aes_type = {
+
+    mp_obj_aesgcm_t *o = m_new_obj(mp_obj_aesgcm_t);
+
+    o->base.type = type;
+    mp_buffer_info_t keyinfo;
+    mp_get_buffer_raise(args[0], &keyinfo, MP_BUFFER_READ);
+    if (32 != keyinfo.len && 16 != keyinfo.len) {
+        mp_raise_ValueError(MP_ERROR_TEXT("key length"));
+    }
+
+    o->key_len = keyinfo.len;
+    memcpy(&o->key, keyinfo.buf, keyinfo.len);
+
+    return MP_OBJ_FROM_PTR(o);
+}
+
+STATIC mp_obj_t ucryptolib_aesgcm_enc_dec_check_args(size_t n_args, const mp_obj_t *args, mp_map_t *kwargs,
+    mp_buffer_info_t *nonce, mp_buffer_info_t *input, mp_buffer_info_t *output, mp_buffer_info_t *ad, int in_out_delta) {
+    mp_obj_t out_obj;
+    mp_map_elem_t *map_element;
+
+    if (n_args != 3) {
+        mp_raise_ValueError(MP_ERROR_TEXT("2 positional arguments required"));
+    }
+
+    mp_get_buffer_raise(args[1], nonce, MP_BUFFER_READ);
+    mp_get_buffer_raise(args[2], input, MP_BUFFER_READ);
+
+    size_t output_len = input->len + in_out_delta;
+
+    map_element = mp_map_lookup(kwargs, MP_OBJ_NEW_QSTR(MP_QSTR_out_buf), MP_MAP_LOOKUP);
+    if (map_element != NULL) {
+        out_obj = map_element->value;
+        mp_get_buffer_raise(out_obj, output, MP_BUFFER_WRITE);
+        if (output->len != output_len) {
+            mp_raise_ValueError(MP_ERROR_TEXT("incorrect output length"));
+        }
+    } else {
+        out_obj = MP_OBJ_NULL;
+        output->buf = m_new(char, output_len);
+        output->len = output_len;
+    }
+
+    map_element = mp_map_lookup(kwargs, MP_OBJ_NEW_QSTR(MP_QSTR_adata), MP_MAP_LOOKUP);
+    if (map_element != NULL) {
+        mp_get_buffer_raise(map_element->value, ad, MP_BUFFER_READ);
+    } else {
+        ad->buf = NULL;
+        ad->len = 0;
+    }
+
+    // Maybe this should check for spurious keyword args...
+
+    return out_obj;
+}
+
+#define UCRYPTOLIB_AESGCM_TAG_LEN 16
+
+STATIC mp_obj_t ucryptolib_aesgcm_encrypt(size_t n_args, const mp_obj_t *args, mp_map_t *kwargs) {
+    mp_obj_aesgcm_t *self = MP_OBJ_TO_PTR(args[0]);
+    mp_buffer_info_t nonce;
+    mp_buffer_info_t input;
+    mp_buffer_info_t output;
+    mp_obj_t out_obj;
+    mp_buffer_info_t ad;
+
+    out_obj = ucryptolib_aesgcm_enc_dec_check_args(n_args, args, kwargs,
+        &nonce, &input, &output, &ad,
+        UCRYPTOLIB_AESGCM_TAG_LEN);
+
+    mbedtls_gcm_context gcm;
+    int rc;
+
+    mbedtls_gcm_init(&gcm);
+
+    rc = mbedtls_gcm_setkey(&gcm, MBEDTLS_CIPHER_ID_AES, self->key, self->key_len * 8);
+
+    int plaintext_len = input.len;
+
+    rc = mbedtls_gcm_crypt_and_tag(&gcm, MBEDTLS_GCM_ENCRYPT,
+        plaintext_len,
+        nonce.buf, nonce.len,
+        ad.buf, ad.len,
+        input.buf, output.buf,
+        UCRYPTOLIB_AESGCM_TAG_LEN,
+        ((unsigned char *)output.buf) + plaintext_len);
+
+    mbedtls_gcm_free(&gcm);
+
+    if (rc != 0) {
+        mp_raise_OSError(rc);
+    }
+
+    if (out_obj == MP_OBJ_NULL) {
+        out_obj = mp_obj_new_str_copy(&mp_type_bytes, output.buf, output.len);
```

Comment:
Actually the problem with doing that is that since strings and bytes are immutable the creation functions compute the hash when the object is created. If I create the object before encrypt/decrypt operation then the empty buffer gets hashed (and then I write over something that is supposed to be immutable). I could find no sign of a function that let me take an existing buffer and hand over ownership to a bytes object, hence doing it this way. If you have a suggestion for a better way to do this which ends up with the correct object hash then Iid be delighted to use it.

Reply:
I think `mp_obj_new_str_from_vstr` can be used for that. It will try to reuse the buffer from the `vstr` object. But you have to make sure the `vstr` is properly allocated with exact size (of `len+1`, + 1 is because all `str`/`bytes` in MicroPython are null-terminated)

Reply:
Thanks. I've switched over to using that.